// @ndn/fw/lib/tracer_node.js
import { Data, Interest, Nack } from "@ndn/packet";
import { console } from "@ndn/util";
import { Forwarder } from "./forwarder_node.js";
/** Print trace logs from Forwarder events. */
export class Tracer {
    constructor({ output = console, fw = Forwarder.getDefault(), face = true, prefix = true, ann = true, pkt = true, }) {
        this.faceadd = (face) => {
            this.output.log(`+Face ${face}`);
        };
        this.facerm = (face) => {
            this.output.log(`-Face ${face}`);
        };
        this.prefixadd = (face, prefix) => {
            this.output.log(`${face} +Prefix ${prefix}`);
        };
        this.prefixrm = (face, prefix) => {
            this.output.log(`${face} -Prefix ${prefix}`);
        };
        this.annadd = (name) => {
            this.output.log(`+Announcement ${name}`);
        };
        this.annrm = (name) => {
            this.output.log(`-Announcement ${name}`);
        };
        this.pktrx = (face, pkt) => {
            this.pkt(face, pkt, ">");
        };
        this.pkttx = (face, pkt) => {
            this.pkt(face, pkt, "<");
        };
        this.output = output;
        this.fw = fw;
        /* istanbul ignore else */
        if (face) {
            this.fw.on("faceadd", this.faceadd);
            this.fw.on("facerm", this.facerm);
        }
        /* istanbul ignore else */
        if (prefix) {
            this.fw.on("prefixadd", this.prefixadd);
            this.fw.on("prefixrm", this.prefixrm);
        }
        /* istanbul ignore else */
        if (ann) {
            this.fw.on("annadd", this.annadd);
            this.fw.on("annrm", this.annrm);
        }
        /* istanbul ignore else */
        if (pkt) {
            this.fw.on("pktrx", this.pktrx);
            this.fw.on("pkttx", this.pkttx);
        }
    }
    static enable(opts = {}) {
        return new Tracer(opts);
    }
    disable() {
        this.fw.off("faceadd", this.faceadd);
        this.fw.off("facerm", this.facerm);
        this.fw.off("prefixadd", this.prefixadd);
        this.fw.off("prefixrm", this.prefixrm);
        this.fw.off("annadd", this.annadd);
        this.fw.off("annrm", this.annrm);
        this.fw.off("pktrx", this.pktrx);
        this.fw.off("pkttx", this.pkttx);
    }
    pkt(face, pkt, dir) {
        switch (true) {
            case pkt.l3 instanceof Interest: {
                const act = pkt.cancel ? "Cancel" :
                    pkt.reject ? `Reject(${pkt.reject})` :
                        "I";
                this.output.log(`${face} ${dir}${act} ${interestToString(pkt.l3)}`);
                break;
            }
            case pkt.l3 instanceof Data: {
                const { name } = pkt.l3;
                this.output.log(`${face} ${dir}D ${name}`);
                break;
            }
            case pkt.l3 instanceof Nack: {
                const { interest, reason } = pkt.l3;
                this.output.log(`${face} ${dir}N ${interestToString(interest)}~${reason}`);
                break;
            }
        }
    }
}
function interestToString({ name, canBePrefix, mustBeFresh }) {
    return `${name}${canBePrefix ? "[P]" : ""}${mustBeFresh ? "[F]" : ""}`;
}
